File: algorithms/java/src/test/java/leetcode/LC096UniqueBinarySearchTrees.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/unique-binary-search-trees/
 */
public class LC096UniqueBinarySearchTrees {
    public class Solution {
        public int numTrees(int n) {
            int[] cache = new int[n + 1];
            cache[0] = cache[1] = 1;
            for (int i = 2; i <= n; i++) {
                for (int j = 0; j < i; j++) {
                    cache[i] += cache[j] * cache[i - j - 1];
                }
            }
            return cache[n];
        }
    }

    @Test
    public void test_2() throws Exception {
        assertEquals(2, new Solution().numTrees(2));
    }

    @Test
    public void test_3() throws Exception {
        assertEquals(5, new Solution().numTrees(3));
    }

    @Test
    public void test_19() throws Exception {
        assertEquals(1767263190, new Solution().numTrees(19));
    }
}

File: algorithms/code/leetcode/lc096_unique_binary_search_trees/lc096_unique_binary_search_trees.py
#!/usr/bin/env python3
# https://leetcode.com/problems/unique-binary-search-trees/

import unittest


class Solution:
    def numTrees(self, n: int) -> int:
        cache = [0] * (n + 1)
        cache[0] = 1
        cache[1] = 1
        for i in range(2, n + 1):
            for j in range(i):
                cache[i] += cache[j] * cache[i - j - 1]
        return cache[n]


class TestCode(unittest.TestCase):
    def test_2(self):
        self.assertEqual(Solution().numTrees(2), 2)

    def test_3(self):
        self.assertEqual(Solution().numTrees(3), 5)

    def test_19(self):
        self.assertEqual(Solution().numTrees(19), 1767263190)
